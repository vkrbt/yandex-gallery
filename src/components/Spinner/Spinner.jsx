import React from 'react';

const Spinner = () => (
  <div className="spinner-wrapper">
    <div className="spinner" />
  </div>
);

export default Spinner;
