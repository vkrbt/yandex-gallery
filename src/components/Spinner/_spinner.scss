@keyframes spin {
  from {
    transform: rotateZ(0);
  }
  to {
    transform: rotateZ(360deg);
  }
}

.spinner-wrapper {
  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  height: 200px;
}

.spinner {
  width: 40px;
  height: 40px;

  animation: spin 1s linear infinite;

  border: 2px solid #555;
  border-top-color: transparent;
  border-radius: 50%;
}
