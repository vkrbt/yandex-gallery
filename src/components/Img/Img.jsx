import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '../Spinner/Spinner';

export class Img extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoaded: false,
      error: null,
    };

    this.handleImageLoaded = this.handleImageLoaded.bind(this);
    this.handleImageError = this.handleImageError.bind(this);
    this.loadImage = this.loadImage.bind(this);
  }

  componentDidMount() {
    this.loadImage(this.props.src);
  }

  loadImage(src) {
    this.setState({ isLoaded: false });
    this.image = new Image();
    this.image.addEventListener('load', this.handleImageLoaded);
    this.image.addEventListener('error', this.handleImageError);
    this.image.src = src;
  }

  handleImageLoaded() {
    this.setState({
      isLoaded: true,
    });
  }

  handleImageError(error) {
    this.setState({ error });
  }

  render() {
    if (this.state.error) {
      return null;
    }
    return this.state.isLoaded ? (
      <img className={this.props.className} src={this.image.src} alt={this.props.alt} />
    ) : (
      <Spinner loading={!this.state.isLoaded} text="Retry" action={this.loadImage} />
    );
  }
}

Img.defaultProps = {
  alt: '',
  className: '',
};

Img.propTypes = {
  src: PropTypes.string.isRequired,
  className: PropTypes.string,
  alt: PropTypes.string,
};
