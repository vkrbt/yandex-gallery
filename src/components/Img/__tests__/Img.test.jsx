import React from 'react';
import { shallow } from 'enzyme';
import { Img } from '../Img';
import { Spinner } from '../../Spinner/Spinner';

describe('Img', () => {
  it('should being Spinner until image not loaded', () => {
    const wrapper = shallow(<Img src="some url" />);
    wrapper.setState({ isLoaded: false, error: null });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Spinner)).toHaveLength(1);
  });

  it('should not being rendered if error was occured', () => {
    const wrapper = shallow(<Img src="some url" />);
    wrapper.setState({ error: true });
    expect(wrapper.html()).toBe(null);
  });

  it('should render image if image is loaded.', () => {
    const wrapper = shallow(<Img src="some url" />);
    wrapper.setState({ isLoaded: true, error: null });
    expect(wrapper.find('img')).toHaveLength(1);
  });
});
