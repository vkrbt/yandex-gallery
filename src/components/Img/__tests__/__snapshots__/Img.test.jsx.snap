// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Img should being Spinner until image not loaded 1`] = `
<Component
  action={[Function]}
  loading={true}
  text="Retry"
/>
`;
